gobblin-modules/google-ingestion/src/main/java/gobblin/ingestion/google/webmaster/GoogleWebmasterDataFetcherImpl.java

```diff
@@ -0,0 +1,366 @@
+      LOG.info(String.format("A total of %d pages fetched for property %s at country-%s from %s to %s", actualSize,
+          _siteProperty, country, startDate, endDate));
+    } else {
+      int expectedSize = getPagesSize(startDate, endDate, country, requestedDimensions, Arrays.asList(countryFilter));
+      LOG.info(String.format("Total number of pages is %d for market-%s from %s to %s", expectedSize,
+          GoogleWebmasterFilter.countryFilterToString(countryFilter), startDate, endDate));
+      Queue<Pair<String, FilterOperator>> jobs = new ArrayDeque<>();
+      expandJobs(jobs, _siteProperty);
+
+      allPages = getPages(startDate, endDate, requestedDimensions, countryFilter, jobs);
+      allPages.add(_siteProperty);
+      actualSize = allPages.size();
+      if (actualSize != expectedSize) {
+        LOG.warn(
+            String.format("Expected page size for country-%s is %d, but only able to get %d", country, expectedSize,
+                actualSize));
+      }
+      LOG.info(String.format("A total of %d pages fetched for property %s at country-%s from %s to %s", actualSize,
+          _siteProperty, country, startDate, endDate));
+    }
+
+    ArrayDeque<ProducerJob> jobs = new ArrayDeque<>(actualSize);
+    for (String page : allPages) {
+      jobs.add(new SimpleProducerJob(page, startDate, endDate));
+    }
+    return jobs;
+  }
+
+  private int getPagesSize(final String startDate, final String endDate, final String country,
+      final List<Dimension> requestedDimensions, final List<ApiDimensionFilter> apiDimensionFilters)
+      throws IOException {
+
+    int startRow = 0;
+    int count = 4;
+    final ExecutorService es = Executors.newFixedThreadPool(count);
+    int r = 0;
+    while (r < 50) {
+      r++;
+      List<Future<Integer>> results = new ArrayList<>(count);
+
+      for (int i = 0; i < count; ++i) {
+        startRow += GoogleWebmasterClient.API_ROW_LIMIT;
+        final int start = startRow;
+        Future<Integer> submit = es.submit(new Callable<Integer>() {
+          @Override
+          public Integer call() {
+            LOG.info(String.format("Getting page size from %s...", start));
+            while (true) {
+              try {
+                List<String> pages =
+                    _client.getPages(_siteProperty, startDate, endDate, country, GoogleWebmasterClient.API_ROW_LIMIT,
+                        requestedDimensions, apiDimensionFilters, start);
+                if (pages.size() < GoogleWebmasterClient.API_ROW_LIMIT) {
+                  //Figured out the size
+                  return pages.size() + start;
+                } else {
+                  //The size is not determined.
+                  return -1;
+                }
+              } catch (IOException e) {
+                LOG.info(String.format("Getting page size from %s failed. Retrying...", start));
+              }
+              try {
+                Thread.sleep(200);
```

Comment:
After I apply the retryer, I feel like the code becomes a little bit harder to read. Please compare the following.
With retryer:
`List<Future<Integer>> results = new ArrayList<>((int) groupSize);

    Map<String, Object> configMap = ImmutableMap.<String, Object>builder()
        .put(RetryerFactory.RETRY_TIME_OUT_MS, TimeUnit.MINUTES.toMillis(TIME_OUT))
        .put(RetryerFactory.RETRY_INTERVAL_MS, Math.round(1000 / REQUESTS_PER_SECOND))
        .put(RetryerFactory.RETRY_TYPE, RetryerFactory.RetryType.FIXED.name()).build();
    Config retryConfig = ConfigFactory.parseMap(configMap);

    final Retryer<Integer> retryer = RetryerFactory.newInstance(retryConfig);

    while (true) {
      for (int i = 0; i < groupSize; ++i) {
        startRow += GoogleWebmasterClient.API_ROW_LIMIT;
        final int start = startRow;
        final String interruptedMsg = String
            .format("Interrupted while trying to get the size of all pages for %s. Current start row is %d.", country,
                start);

        Future<Integer> submit = es.submit(new Callable<Integer>() {
          @Override
          public Integer call() {
            log.info(String.format("Getting page size from %s...", start));

            Callable<Integer> retryCallable = new Callable<Integer>() {
              @Override
              public Integer call()
                  throws Exception {
                try {
                  LIMITER.acquirePermits(1);
                  List<String> pages = _client
                      .getPages(_siteProperty, startDate, endDate, country, GoogleWebmasterClient.API_ROW_LIMIT,
                          requestedDimensions, apiDimensionFilters, start);
                  if (pages.size() < GoogleWebmasterClient.API_ROW_LIMIT) {
                    return pages.size() + start;  //Figured out the size
                  } else {
                    return -1;
                  }
                } catch (InterruptedException e) {
                  log.error("RateBasedLimiter: " + interruptedMsg, e);
                  return -1;
                }
              }
            };

            try {
              return retryer.call(retryCallable);
            } catch (ExecutionException e) {
              throw new RuntimeException(e);
            } catch (RetryException e) {
              throw new RuntimeException(String
                  .format("Exceeding the timeout of %d minutes while getting the total size of all pages.", TIME_OUT),
                  e);
            }
          }
        });
        results.add(submit);
      }
      //Check the results group in order. The first non-negative count indicates the size of total pages.
      for (Future<Integer> result : results) {
        try {
          Integer integer = result.get();
          if (integer >= 0) {
            es.shutdownNow();
            return integer;
          }
        } catch (InterruptedException | ExecutionException e) {
          throw new RuntimeException(e);
        }
      }
      results.clear();
    }` 

Previous version without retryer:
`List<Future<Integer>> results = new ArrayList<>((int) groupSize);

    while (true) {
      for (int i = 0; i < groupSize; ++i) {
        startRow += GoogleWebmasterClient.API_ROW_LIMIT;
        final int start = startRow;
        final String interruptedMsg = String
            .format("Interrupted while trying to get the size of all pages for %s. Current start row is %d.", country,
                start);

        Future<Integer> submit = es.submit(new Callable<Integer>() {
          @Override
          public Integer call() {
            log.info(String.format("Getting page size from %s...", start));
            while (true) {
              if (Thread.interrupted()) {
                log.error(interruptedMsg);
                return -1;
              }
              try {
                LIMITER.acquirePermits(1);
                List<String> pages = _client
                    .getPages(_siteProperty, startDate, endDate, country, GoogleWebmasterClient.API_ROW_LIMIT,
                        requestedDimensions, apiDimensionFilters, start);
                if (pages.size() < GoogleWebmasterClient.API_ROW_LIMIT) {
                  return pages.size() + start;  //Figured out the size
                } else {
                  return -1;
                }
              } catch (IOException e) {
                log.info(String.format("Getting page size from %s failed. Retrying...", start));
              } catch (InterruptedException e) {
                log.error("RateBasedLimiter: " + interruptedMsg, e);
                return -1;
              }
            }
          }
        });
        results.add(submit);
      }
      //Check the results group in order. The first non-negative count indicates the size of total pages.
      for (Future<Integer> result : results) {
        try {
          Integer integer = result.get(GET_PAGE_SIZE_TIME_OUT, TimeUnit.MINUTES);
          if (integer >= 0) {
            es.shutdownNow();
            return integer;
          }
        } catch (InterruptedException | ExecutionException e) {
          throw new RuntimeException(e);
        } catch (TimeoutException e) {
          throw new RuntimeException(String
              .format("Exceeding the timeout of %d minutes while getting the total size of all pages.",
                  GET_PAGE_SIZE_TIME_OUT), e);
        }
      }
      results.clear();
    }`

I actually feel that the version without retryer is better in this case.

For other cases, I cannot use the retryer because the job can evolve based on the results, but the retryer only supports retrying static jobs.